### ISSUE TITLE
How to train on multi host tpu pod

### ISSUE THREAD

**Issue by siddagra** [google/tunix#560]
How to train on multi host tpu pod like TPUv4-32? It says I need to ssh into every machine but I am confused how it works for Tunix library specifically

**Comment by carlesoctav** [comment]
TLDR: You need a way to command the other machines to run the same script as the main host in one go. One of the easiest ways to do this is by using `tpux`.
Another approach is to use Ray for distributed execution and connect all the machines under a single Docker (I just remembered that vLLM does something similar to host across multiple machines).
See more: [https://docs.jax.dev/en/latest/multi_process.html](https://docs.jax.dev/en/latest/multi_process.html)

**Comment by opooladz** [comment]
would be nice if tunix would just do something like what david  built here and took care of all that stuff for users

https://github.com/marin-community/marin/tree/main

**Comment by tianshub** [comment]
The current multi-host training remains as a preview features in GCP, where underneath it's leveraging Pathways as the single controller backend (similar to what Ray offers but is better abstracted and integrated with Jax). What @carlesoctav mentioned is another viable option to run distributed training with multi-controller, but for RL workflow it can be tricky since the workload on each host can be asymmetric.

Unfortunately the preview feature is currently protected by an allow list, @siddagra if you have a GCP account and TPU resources there, we can figure out how to add you to the list. The good news is, there's an ongoing OSS pathways effort coming up soon, and we will expect that to be much more publicly available (hopefully by the end of Q4 or early Q1 next year).

**Comment by tianshub** [comment]
@opooladz great call, we are actually in touch with David's lab on the Marin effort :) as I mentioned above, we picked Pathways over Ray for distributed setup since it's more naturally integrated with Jax, we are looking forward for the OSS support soon.

**Comment by opooladz** [comment]
Oof I really hope Pathways isn't the only path forward for this project. Pathways on GCP depends on GKE, and GKE is not compatible with TRC. That would lock out a lot of people.

**Comment by siddagra** [comment]
@carlesoctav  Ok so after I have setup tpux, I just run a python script, and the training will happen over the whole pod as long as I define the `MESH` properly?
Also, if I want to use jupyter notebook to train using tpuv4-32, is that possible?

**Comment by siddagra** [comment]
@tianshub Yes I do have a GCP account with TPU but I am using the TRC program.
My project ID is: rl-projects-472617

**Comment by tianshub** [comment]
@opooladz definitely hear your concern on pathways' general availability, that's why I mentioned OSS pathways (which is a separate effort to the existing pathways offering on GCP vis GKE). That will make it much more available for OSS users and there's no cloud lock-in. Think about OSS Pathways as an alternative to Ray, but more native performant on TPU.

**Comment by tianshub** [comment]
@siddagra  ouch..TRC doesn't work with cloud Pathways yet unfortunately (there are ongoing discussion how to make it available since we heard this from a lot of research labs, but it's not concluded yet). What's the training scale you need? if model is not too large we can see if it can fit on a single v5p (96GB HBM per chip * 4) which should be available in TRC pool.

**Comment by siddagra** [comment]
@tianshub Thanks for all the help! From you and everyone else too!
I am trying to experiment with various GRPO improvements in my lab. For now I have recreated some ideas from the following papers:
- [GRPO](https://arxiv.org/abs/2402.03300)
- [GSPO](https://www.arxiv.org/pdf/2507.18071)
- [Magistral](https://arxiv.org/pdf/2506.10910)
- [DR-GRPO](https://arxiv.org/pdf/2503.20783)
- [DAPO (Clip Higher)](https://arxiv.org/pdf/2503.14476)
- [VAPO](https://arxiv.org/abs/2504.05118)

You may find this under the pull request: https://github.com/google/tunix/pull/557 

Ideally I wanted to run Qwen 7B with a batch size of 32 and 16 completions per prompt, similar to the [BNPO Paper](https://arxiv.org/pdf/2506.02864), but I realise that may not be possible, so I was just trying to get a `batch_size` of 8 and `num_generations` to 8 on a 2B param model with LoRA for now. On a single TPUv4-8 slice I am able to run `batch_size = 2` and `num_generations = 8` so I thought with TPUv4-32 which you get on TRC, I can run `batch_size = 8` and `num_generations = 8`.

But the mail I got did not show v5p. Perhaps it is offered but it was not offered to me? Or perhaps it is offered but they are using an older stale automated mail?
> Thanks again for your interest in using Cloud TPUs to accelerate your machine learning research. Your Google Cloud project rl-projects-472617 now has access to the following quota free of charge for 30 days:
> 64 spot Cloud TPU v6e chips in zone europe-west4-a
> 64 spot Cloud TPU v6e chips in zone us-east1-d
> 64 spot Cloud TPU v5e chips in zone us-central1-a
> 64 spot Cloud TPU v5e chips in zone europe-west4-b
> 32 spot Cloud TPU v4 chips in zone us-central2-b
> 32 on-demand Cloud TPU v4 chips in zone us-central2-b

But yeah anything you could do to help get this to run, or speed up experimentation would be helpful!

It is hard to get anything meaningful out of a batch size of 2 currently. Though I am noticing that GSPO consitently underperforms GRPO, and I regularly get large KL spikes around 800-1300 steps in my setup (similar to the default notebook for grpo demo). 

I also reached out to marin as a github issue, if they could allow me to submit some pull requests for RL runs, but I am not sure because they seem to only have documentation for scaling laws speedrun, data mixes, etc.

Thanks for any help/suggestions!

**Comment by siddagra** [comment]
Would it be possible to just use tpux and get this to work across multiple hosts tpu pod?

**Comment by tianshub** [comment]
@siddagra thanks for sharing the additional details and excited to see you already have PR for the algo.

I think for the qwen-7b model it can still fit onto a single v6e-8 host (I think you have it in your TRC pool). Each v6e chip has 32GB HBM so total you have 256GB. Qwen-7b with fp32 for training takes ~112GB HBM for model weights, grad and optimizer states (assuming Adam). Also assuming your generation length is not crazily long. Below is a few things you can do:

1. Enable remat (activation checkpointing) by setting [remat_config to be BLOCK](https://github.com/google/tunix/blob/main/tunix/models/qwen2/model.py#L98). This will largely reduce the activation HBM usage during training.

2. Enable [CPU offloading](https://github.com/google/tunix/blob/main/tunix/rl/rl_cluster.py#L177), this will offload models at each stage to CPU and free up the HBM.

3. Leverage [micro_batching](https://github.com/google/tunix/blob/main/tunix/rl/rl_cluster.py#L99-L104), which can break down the global batch into smaller ones and avoid OOM issues. On trainer side gradient accumulation will be automatically enabled to handle the micro batches.

4. Try with LoRA layers and share the base model weights between rollout and actor (this way you don't need CPU offloading).

Let us know if the above options work (since it's tricky to enable multi-host TPU without pathways right away, and trust me the team is working on adding GKE support in TRC :) )

**Comment by opooladz** [comment]
That's not a good solution imo.

Why not just use something like eopod to distribute over different servers. The issue isn't really distribution. I think the issue is being robust to preemption which is required for spot compute on TRC.

**Comment by srikilaru-google** [comment]
@siddagra @opooladz 
Could you please use this form to help us better understand your background, interest and needs -
https://docs.google.com/forms/d/e/1FAIpQLSfg6hsk6X3dPFKveR7DGDS7sHL8qQV6jgFsTZVH4VFQCfraFw/viewform

**Comment by siddagra** [comment]
@srikilaru-google I submitted the form. Thanks!

I would also love to actively contribute to this repository through my experiments/tests as I am currently implementing many popular GRPO variants (Dr. GRPO, BNPO, Magistral, GRPO-$lambda$, etc.,) as a baseline. If that can help me get this running on multi-host TPU or help me get more resources, I would be grateful and would love to actively contribute.

As previously stated by me:
> I am trying to experiment with various GRPO improvements in my lab. For now I have recreated some ideas from the following papers:
> 
> * [GRPO](https://arxiv.org/abs/2402.03300)
> * [GSPO](https://www.arxiv.org/pdf/2507.18071)
> * [Magistral](https://arxiv.org/pdf/2506.10910)
> * [DR-GRPO](https://arxiv.org/pdf/2503.20783)
> * [DAPO (Clip Higher)](https://arxiv.org/pdf/2503.14476)
> * [VAPO](https://arxiv.org/abs/2504.05118)
> 
> You may find this under the pull request: [#557](https://github.com/google/tunix/pull/557)

**Comment by siddagra** [comment]
> That's not a good solution imo.
> 
> Why not just use something like eopod to distribute over different servers. The issue isn't really distribution. I think the issue is being robust to preemption which is required for spot compute on TRC.

@opooladz 

Would it help if I have access to on demand non-spot TPUs access?

**Comment by opooladz** [comment]
I mean if you are on TRC you can't use pathways. Doesn't matter for spot or on demand

**Comment by nicolomonti** [comment]
Hey, I’ve been using Tunix with Ray successfully. The process is pretty straightforward, but some parts (e.g., RL) require extra attention to work well on multi-host setups. If you need to use it right now, I recommend taking something like EasyDeL’s Ray autoscaler setup and adapting it to Tunix, although it’s undoubtedly a bit of a hacky solution.

**Comment by siddagra** [comment]
@nicolomonti Can u tell me what the step by step process is. I am not too familiar with TPUs and the documentation/LLM knowledge on it also seems scarce for me to do this stuff without some guides. I would be very grateful if you could share  how to get this to work on multi host TPUs. 

I was hoping I could just use tpux and then run this code from all workers and it would run in parallel properly?

Not to be ungrateful cuz I am very grateful for free TPUs from TRC. But this multi-host SSH stuff is the most scuffed thing I have ever seen and makes me want to switch to some free GPU-based resources.

**Comment by wang2yn84** [comment]
Hi @nicolomonti, would you like to raise a PR to add Ray support for Tunix?

**Comment by nicolomonti** [comment]
Hey! I'll definitely submit a PR by Monday. However, I won't include multi-host support, as this would involve pushing several potentially breaking changes.

**Comment by yhtang** [comment]
Here is a related issue: #698

**Comment by wang2yn84** [comment]
Hi everyone, please take a look at the [RFC](https://github.com/google/tunix/discussions/706) for McJax support and feedbacks are more than welcome!

**Comment by ahmeda14960** [comment]
@tianshub Following up on this thread please, is there a plan to make this pathways support OSS soon? If not could we set up a preview somehow?

**Comment by tianshub** [comment]
@ahmeda14960 pathways will be available on GKE right now, outside of that, we are forming an internal plan and will share once it's taken better shape.
